File: MatxVec.c
#include <stdio.h>
#include <stdlib.h>


void mxvSeq(int m, int n, double * restrict a,
        double * restrict b, double * restrict c)
{
    int i, j;

    for (i=0; i<m; i++) {
        a[i] = 0.0;
        for (j=0; j<n; j++) {
            a[i] += b[i*n+j] * c[j];
        }
    }
}


void mxvOpenMP(int m, int n, double * restrict a,
        double * restrict b, double * restrict c)
{
    int i, j;

    #pragma omp parallel for default(none) shared(m,n,a,b,c) private(i,j)
    for (i=0; i<m; i++) {
        a[i] = 0.0;
        for (j=0; j<n; j++) {
            a[i] += b[i*n+j] * c[j];
        }
    }
}


int main(int argc, char *argv[])
{
    double *a, *b, *c;

    int i, j;
    int m, n;

    if (argc < 3) {
        printf("Please enter m and n: ");
        scanf("%d %d", &m, &n);
    } else {
        m = atoi(argv[1]);
        n = atoi(argv[2]);
    }

    a = (double *)malloc(m * sizeof(double));
    if (a == NULL) {
        perror("memory allocation for a");
    }
    b = (double *)malloc(m * n * sizeof(double));
    if (b == NULL) {
        perror("memory allocation for b");
    }
    c = (double *)malloc(n * sizeof(double));
    if (c == NULL) {
        perror("memory allocation for c");
    }


    printf("Initializing matrix B and vector C\n");
    for (i=0; i<m; i++) {
        for (j=0; j<n; j++) {
            b[i*n+j] = 1.0;
        }
    }
    for (j=0; j<n; j++) {
        c[j] = 2.0;
    }


    printf("Executing mxv function for m=%d, n=%d\n", m, n);
    //mxvSeq(m, n, a, b, c);
    mxvOpenMP(m, n, a, b, c);

    //for (i=0; i<m; i++) {
    //    printf("%7.2g  ", a[i]);
    //}
    //printf("\n");

    free(a);
    free(b);
    free(c);

    return 0;
}

File: MatxVec.f
module mxv
    implicit none

    contains

    subroutine mxvSeq(m, n, a, b, c)
        implicit none
        integer, intent(in)     :: m, n
        real,    intent(in)     :: b(1:m,1:n), c(1:n)
        real,    intent(inout)  :: a(1:m)

        integer :: i, j

        do i=1, m
            a(i) = 0.0
            do j=1, n
                a(i) = a(i) + b(i,j) * c(j)
            end do
        end do
    end subroutine

    subroutine mxvOpenMP(m, n, a, b, c)
        implicit none
        integer, intent(in)     :: m, n
        real,    intent(in)     :: b(1:m,1:n), c(1:n)
        real,    intent(inout)  :: a(1:m)

        integer :: i, j

        !$omp parallel do default(none) shared(m,n,a,b,c) private(i,j)
        do i=1, m
            a(i) = 0.0
            do j=1, n
                a(i) = a(i) + b(i,j) * c(j)
            end do
        end do
        !$omp end parallel do
    end subroutine
end module


program main
    use mxv
    implicit none

    real, allocatable :: a(:), b(:,:), c(:)
    integer :: m, n, i, memstat

    write (*,*) "Please give me m and n:"
    read (*,*) m, n

    allocate(a(1:m), stat=memstat)
    if (memstat .ne. 0) then
        stop "Error in memory allocation for 'a'"
    end if

    allocate(b(1:m, 1:n), stat=memstat)
    if (memstat .ne. 0) then
        stop "Error in memory allocation for 'b'"
    end if

    allocate(c(1:n), stat=memstat)
    if (memstat .ne. 0) then
        stop "Error in memory allocation for 'c'"
    end if

    ! initialize matrix B and vector C
    c(1:n) = 1.0
    do i=1, m
        b(i, 1:n) = i
    end do

    call mxvSeq(m, n, a, b, c)

    if (allocated(a)) deallocate(a, stat=memstat)
    if (allocated(b)) deallocate(b, stat=memstat)
    if (allocated(c)) deallocate(c, stat=memstat)
end program
